### ISSUE TITLE
Loss is constant, 10 and not decreasing

### ISSUE THREAD

**Issue by adesgautam** [google/automl#834]
### I am training the model on custom dataset and i am getting this below:

INFO:tensorflow:loss = 10.092902, step = 10
I1023 08:59:37.865594 140110336001920 basic_session_run_hooks.py:262] loss = 10.092902, step = 10
INFO:tensorflow:box_loss = 0.0, cls_loss = 10.0, det_loss = 10.0, step = 10
I1023 08:59:37.866737 140110336001920 basic_session_run_hooks.py:262] box_loss = 0.0, cls_loss = 10.0, det_loss = 10.0, step = 10
INFO:tensorflow:Calling checkpoint listeners before saving checkpoint 30...
I1023 09:00:18.539453 140110336001920 basic_session_run_hooks.py:614] Calling checkpoint listeners before saving checkpoint 30...
INFO:tensorflow:Saving checkpoints for 30 into efficientdet-d0-finetune/model.ckpt.
I1023 09:00:18.539808 140110336001920 basic_session_run_hooks.py:618] Saving checkpoints for 30 into efficientdet-d0-finetune/model.ckpt.
INFO:tensorflow:Calling checkpoint listeners after saving checkpoint 30...
I1023 09:00:23.955784 140110336001920 basic_session_run_hooks.py:626] Calling checkpoint listeners after saving checkpoint 30...
INFO:tensorflow:Loss for final step: 10.092899.
I1023 09:00:24.551039 140110336001920 estimator.py:350] Loss for final step: 10.092899.

The loss is constant and not getting reduced. I am doing this on Colab.
Any help is appreciated.

### config.yaml :
num_classes: 5
var_freeze_expr: '(efficientnet|fpn_cells|resample_p6)'
label_map: {1: cls1, 2: cls2, 3: cls3, 4: cls4}
learning_rate: 0.001
lr_warmup_init: 0.0001

**Comment by NikZak** [comment]
Try to not freeze the layers
```
num_classes: 5
# var_freeze_expr: '(efficientnet|fpn_cells|resample_p6)'
label_map: {1: cls1, 2: cls2, 3: cls3, 4: cls4}
learning_rate: 0.001
lr_warmup_init: 0.000
```

**Comment by adesgautam** [comment]
Tried this, still the same

**Comment by MiJiw** [comment]
Hi, I have the same issue with latest code (and the one from early october), both with Keras and legacy with tf2. Cls_loss is stuck at 10.0. 
I think it comes from sgd. Tested against several value of lr rate, with and without freezing anything. I have examples where the loss jump back to normal after several thousand iterations.
What is strange it's that is highly dependant of the custom dataset used. Some always work, some work with no augmentation, some don't work no matter what is tried. 

One workaround is to use adam as the optimizer. With 100+ training never had problems with it but results can sometimes be impacted compared to sgd (on dataset that can accept both optimizer).

If someone can point out why and where is it capped at 10 (didn't find it myself) maybe we could find a solution and test on multiple custom dataset?
@fsx950223, @mingxingtan, maybe you could give a hand?

Feel free to ask whatever details you may need. Thanks for your help.

![custom_effdet-d1](https://user-images.githubusercontent.com/50751053/98938941-f668ad80-24e8-11eb-92d3-862647529e22.png)

**Comment by fsx950223** [comment]
https://github.com/google/automl/blob/0ebd3d0a468180d52529c6a3844ecfc67b0a0fbe/efficientdet/det_model_fn.py#L398
Maybe you should tune ```clip_gradients_norm``` up.

**Comment by MiJiw** [comment]
It was my first idea back in october, I tried with clip_gradients_norm: 100.0 but it didn't worked. 
I could try larger values with latest code and report back.

**Comment by fsx950223** [comment]
https://github.com/google/automl/blob/0ebd3d0a468180d52529c6a3844ecfc67b0a0fbe/efficientdet/det_model_fn.py#L260

**Comment by MiJiw** [comment]
I checked the graph of gradient_norm of a large number of problematic run, it doesn't seem to be correlated. Am I missing something? 
![clip_grad](https://user-images.githubusercontent.com/50751053/98976826-1cf20d00-2518-11eb-8196-830d41b996b6.png)
![train_loss](https://user-images.githubusercontent.com/50751053/98976833-1f546700-2518-11eb-8b53-f0cd10a68a45.png)

**Comment by thhung** [comment]
I also have the same problem when training with custom dataset. cls_loss always is 10.0. And when you use train_and_eval, the eval is way slower than training which is weird. It seems the model doesn't learn anything.

**Comment by mingxingtan** [comment]
> https://github.com/google/automl/blob/0ebd3d0a468180d52529c6a3844ecfc67b0a0fbe/efficientdet/det_model_fn.py#L260
> 
> Is the sentence necessary? As gradients have been clipped. @mingxingtan

**Comment by mingxingtan** [comment]
@fsx950223 it was a mistake to clip the loss. I have just submitted a fix.

**Comment by MiJiw** [comment]
@mingxingtan Thank you. I'll test the fix and report back.

**Comment by thhung** [comment]
I confirm that for my case, with the new fix, cls_loss is no longer 10.0. AP starts to have something greater than 0.0 . Eval still is very slow comparing with training.

**Comment by MiJiw** [comment]
I also confirm. We can now train with sgd on custom dataset without problem.
Thank you!
![trainloss_ok](https://user-images.githubusercontent.com/50751053/99045823-346ddc00-2592-11eb-9d49-af4a067823b1.png)

**Comment by thhung** [comment]
@MiJiw Would you mind sharing how do you prepare your dataset and command you run for training the model? Currently I trained on a custom dataset. However, even after 32 epochs, my AP is still under 0.01. I would not expect it is this low. The only thing I make differently is the split the original training into a new train and val set before creating the tfrecords. Thanks in advance.

**Comment by MiJiw** [comment]
@thhung Sure!
For the dataset preparation, I also split my training set and follow the steps in the README for creating tfrecords. (I automated that part in 3 subprocess call for the 3 sets but it's the same cmd). If you have doubt on your dataset, it's always a good idea to visualize your annotations and check for negative bbox area or out of frame coordinates.

As for the training, I found that finetuning with variables freezed always deteriorate performance (or like you, give near 0 AP) on my custom datasets. 
It is generally enough for me to train for 8 to 12 epochs with default lr. In that setting, I only change lr_warmup_epoch to 0.2.
If you still have 0 mAP, try training from scratch. If it is still bad, it means you have a problem with your data.

For the slow eval, I also noticed that. I found that it comes from the nms, after changing iou_thresh to 0.5 (soft_nms is slower by nature) and score_thresh to 0.05 (default value for mmdetection for exemple), eval is lightining fast. You may loose mAP decimals on your val/test but if you favor real use case to benchmarks, it's not a problem.

Here is the list of hparams that I tweak the most :
jitter_min: 0.5 is my default value for multiscale training, found it more suitable to my usecases
jitter_max: 1.5
autoaugment_policy: None or 'v3' as it gives better results on multiple domain (visible, thermal and synth)
anchor_scale: 3.0 or 2.0 give generally higher mAP small for me
is_training_bn: Found that you can gain 1-2 mAP by setting it to False on some cases
optimizer: I always try adam and sgd. I also tried others.
learning_rate: default is good start, depends on the dataset. Good to know that it is adjusted in the code wrt your batch_size so no need to divide it yourself.
lr_warmup_init: default or lower
lr_warmup_epoch: 0.2 is my default
num_epochs: 8 to 12 for finetuning
gamma: 2 or more, it adjusts the focal loss. Depends on the granularity of my class mapping.
weight_decay: default but may need tweaking if you try new optimizer like adabelief
lr_decay_method: 'cosine' is good start but 'stepwise' with a last epoch drop could give you better results.

Hope it helps. Good luck!

**Comment by adesgautam** [comment]
I have used this fix: ba5131a and now the loss is going down 10. But there is nothing at inferencing. There are no bounding boxes or anything. The metrics after 10 epochs is below:

I1118 05:59:56.155458 139921240348544 basic_session_run_hooks.py:260] loss = 0.09515423, step = 100 (108.665 sec)
INFO:tensorflow:box_loss = 0.0, cls_loss = 0.0014172639, det_loss = 0.0014172639, step = 100 (108.665 sec)
I1118 05:59:56.155850 139921240348544 basic_session_run_hooks.py:260] box_loss = 0.0, cls_loss = 0.0014172639, det_loss = 0.0014172639, step = 100 (108.665 sec)
INFO:tensorflow:Calling checkpoint listeners before saving checkpoint 160...
